src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RuntimeMetrics.cs

```diff
@@ -0,0 +1,248 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+#if NET
+using System.Threading;
+#endif
+
+namespace System.Diagnostics.Metrics
+{
+    internal static class RuntimeMetrics
+    {
+        [ThreadStatic] private static bool t_handlingFirstChanceException;
+
+        private const string MeterName = "System.Runtime";
+
+        private static readonly Meter s_meter = new(MeterName);
+
+        // These MUST align to the possible attribute values defined in the semantic conventions (TODO: link to the spec)
+        private static readonly string[] s_genNames = ["gen0", "gen1", "gen2", "loh", "poh"];
+#if NET
+        private static readonly int s_maxGenerations = Math.Min(GC.GetGCMemoryInfo().GenerationInfo.Length, s_genNames.Length);
+#endif
+
+        static RuntimeMetrics()
+        {
+            AppDomain.CurrentDomain.FirstChanceException += (source, e) =>
+            {
+                if (t_handlingFirstChanceException) return;
```

Comment:
could you add a comment on why this check is required?

Reply:
Added a comment.